ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,246 @@ type clientHandshakeState struct {
 	session         *ClientSessionState
 }
 
+type CacheKeyGenerator interface {
+	Key(net.Addr) string
+}
+
+type ClientHelloConfiguration struct {
+	//Version in the handshake header
+	HandshakeVersion uint16
+
+	//if len == 32, it will specify the client random
+	//Otherwise, the field will be random
+	ClientRandom []byte
+
+	//if RandomSessionID > 0, will overwrite SessionID w/ that many
+	//random bytes when a session resumption occurs
+	RandomSessionID int
+	SessionID       []byte
+
+	//These fields will appear exactly in order in the ClientHello
+	CipherSuites       []uint16
+	CompressionMethods []uint8
+	Extensions         []ClientExtension
+
+	//Optional, both must be non-nil, or neither.
+	//Custom Session cache implementations allowed
+	SessionCache ClientSessionCache
+	CacheKey     CacheKeyGenerator
+}
+
+func (c *ClientHelloConfiguration) ValidateExtensions() error {
+	for _, ext := range c.Extensions {
+		switch ext.(type) {
+		case PointFormatExtension:
+			for _, format := range ext.(PointFormatExtension).Formats {
+				if format != pointFormatUncompressed {
+					return errors.New(fmt.Sprintf("Unsupported EC Point Format %d", format))
+				}
+			}
+		case SignatureAlgorithmExtension:
+			for _, algs := range ext.(SignatureAlgorithmExtension).getStruct() {
+				found := false
+				for _, supported := range supportedSKXSignatureAlgorithms {
+					if algs.hash == supported.hash && algs.signature == supported.signature {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return errors.New(fmt.Sprintf("Unsupported Hash and Signature Algorithm (%d, %d)", algs.hash, algs.signature))
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (c *ClientHelloConfiguration) ModifyConfig(config *Config) *Config {
+	config.NextProtos = []string{}
+	config.CipherSuites = c.CipherSuites
+	config.MaxVersion = c.HandshakeVersion
+	config.ClientRandom = c.ClientRandom
+	config.CurvePreferences = []CurveID{}
+	config.HeartbeatEnabled = false
+	config.ExtendedRandom = false
+	config.ForceSessionTicketExt = false
+	config.ExtendedMasterSecret = false
+	config.SignedCertificateTimestampExt = false
```

Comment:
Blegh. This is the downside of having two separate codepaths. I'm somewhat inclined to say screw it, and remove the original codepath so that we only build ClientHello's using the ClientFingerprint approach, but that also makes me nervous for obvious reasons.

My original worry here was that we would add a field to `Config`, and then forget to update this function. Thoughts?